src/MIDebugEngine/Engine.Impl/Disassembly.cs

```diff
@@ -305,17 +305,153 @@ private void DeleteRangeFromCache(DisassemblyBlock block)
+                 *              address="0x0000000008001485",
+                 *              func-name="main(int, char**)",
+                 *              offset="316",
+                 *              opcodes="83 bd 4c ff ff ff 00",
+                 *              inst="cmpl   $0x0,-0xb4(%rbp)"
+                 *          },
+                 *          {
+                 *              address="0x000000000800148c",
+                 *              func-name="main(int, char**)",
+                 *              offset="323",
+                 *              opcodes="75 07",
+                 *              inst="jne    0x8001495 <main(int, char**)+332>"
+                 *          }
+                 *      ]
+                 *   }
+                 *  ]
+                 */
+                foreach (TupleValue src_and_asm_line in src_and_asm_lines)
+                {
+                    uint line = src_and_asm_line.FindUint("line");
+                    string file = process.GetMappedFileFromTuple(src_and_asm_line);
+                    ValueListValue line_asm_instructions = src_and_asm_line.Find<ValueListValue>("line_asm_insn");
+                    foreach (ResultValue line_asm_insn in line_asm_instructions.Content)
+                    {
+                        ulong address = line_asm_insn.FindAddr("address");
+                        AddressToInstruction[address].File = file;
+                        AddressToInstruction[address].Line = line;
+                    }
+                }
+            }
+        }
+
         // this is inefficient so we try and grab everything in one gulp
         internal static async Task<DisasmInstruction[]> Disassemble(DebuggedProcess process, ulong startAddr, ulong endAddr)
         {
+            // Due to GDB not returning source information when requesting outside of the range of user code.
+            // We first get disassembly with opcodes, then map each Symbol to an address range and attempt to retrieve source information per Symbol.
+
+            // Mode 2 - disassembly with raw opcodes
             string cmd = "-data-disassemble -s " + EngineUtils.AsAddr(startAddr, process.Is64BitArch) + " -e " + EngineUtils.AsAddr(endAddr, process.Is64BitArch) + " -- 2";
             Results results = await process.CmdAsync(cmd, ResultClass.None);
             if (results.ResultClass != ResultClass.done)
             {
                 return null;
             }
 
-            return DecodeDisassemblyInstructions(results.Find<ValueListValue>("asm_insns").AsArray<TupleValue>());
+            DisasmInstruction[] instructions =  DecodeDisassemblyInstructions(results.Find<ValueListValue>("asm_insns").AsArray<TupleValue>());
+
+            if (instructions != null && instructions.Length != 0)
+            {
+                // Map 'Symbol' (Function Name) to Address Range
+                Dictionary<string, DisasmAddressRange> funcToAddressRange = new Dictionary<string, DisasmAddressRange>();
```

Comment:
>Dictionary<string, DisasmAddressRange> [](http://example.com/codeflow?start=16&length=38)

I don't think you want a map here because if you see something like:

MethodA
MethodB
MethodA

You don't want to combine the two MethodA blocks together. So you just need to remember the last (range/function name) pair.